File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
public:
    void exploreIsland(vector<vector<char>> &grid, int i, int j){
        
        if(i<0 || j<0 || i==grid.size() || j== grid[0].size() || grid[i][j]=='0')return;
        
        grid[i][j] = '0';
        exploreIsland(grid, i, j+1);
        exploreIsland(grid, i+1, j);
        exploreIsland(grid, i-1, j);
        exploreIsland(grid, i, j-1);
        
        return;
    }
    
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        for(int i=0; i<grid.size(); i++){
            for(int j=0; j<grid[0].size(); j++){
                if(grid[i][j]=='1'){
                    count++;
                    exploreIsland(grid, i, j);
                }
            }
        }
        return count;
    }
};

File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length, n = grid[0].length;
        int count = 0;
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j]=='1'){
                    count++;
                    bfs(grid, i, j);
                }
            }
        }
        return count;
    }
    private void bfs(char[][] grid, int r, int c){
        int m = grid.length, n = grid[0].length;
        int[] dirs = new int[]{0, 1, 0, -1, 0};
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        grid[r][c] = '0';
        queue.offer(new int[]{r, c});
        while(!queue.isEmpty()){
            int[] cur = queue.poll();
            for(int i=0; i<4; i++){
                int nr = cur[0]+dirs[i];
                int nc = cur[1]+dirs[i+1];
                if(nr<0 || nr>=m || nc<0 || nc>=n || grid[nr][nc]!='1')continue;
                grid[nr][nc] = '0';
                queue.offer(new int[]{nr, nc});
            }
        }
        return;
    }
}